lib/guides_style_18f/navigation.rb

```diff
@@ -23,17 +23,79 @@ def self.validate_with_message_upon_error(front_matter)
 
     def self.validate(front_matter)
       front_matter.map do |file, data|
-        next [file, ['no front matter defined']] unless data.instance_of? Hash
+        next [file, ['no front matter defined']] if data.nil?
         errors = missing_property_errors(data) + permalink_errors(data)
         [file, errors] unless errors.empty?
       end.compact.to_h
     end
 
+    private
+
+    def self.init_site(basedir)
+      Dir.chdir(basedir) do
+        config = SafeYAML.load_file('_config.yml', safe: true)
+        adjust_config_paths(basedir, config)
+        site = Jekyll::Site.new(Jekyll.configuration(config))
+        site.reset
+        site.read
+        site
+      end
+    end
+
+    def self.adjust_config_paths(basedir, config)
+      source = config['source']
+      config['source'] = source.nil? ? basedir : File.join(basedir, source)
+      destination = config['destination']
+      destination = '_site' if destination.nil?
+      config['destination'] = File.join(basedir, destination)
+    end
+
+    def self.site_file_to_front_matter(site)
+      site_pages(site).map do |page|
+        [page.relative_path,  page.data]
+      end.to_h
+    end
+
+    # We're supporting two possible configurations:
+    # 
+    # - a `pages/` directory in which documents appear as part of the regular
+    #   site.pages collection; we have to filter by page.relative_path, and we
+    #   do not assign a permalink so that validation (in a later step) will
+    #   ensure that each page has a permalink assigned
+    #
+    # - an actual `pages` collection, stored in a `_pages` directory; no
+    #   filtering is necessary, and we can reliably set the permalink to
+    #   page.url as a default
+    def self.site_pages(site)
+      pages = site.collections['pages']
+      if pages.nil?
+        site.pages.select do |page|
+          page.relative_path.start_with?('/pages') || page.url == '/'
+        end
+      else
+        pages.docs.each { |page| page.data['permalink'] ||= page.url }
+      end
+    end
+
+    def self.init_file_to_front_matter_map(basedir)
+      file_to_front_matter = {}
+      Dir.chdir(basedir) do
+        pages_dir = Dir.exist?('_pages') ? '_pages' : 'pages'
+        Dir[File.join(pages_dir, '**', '*')].each do |file_name|
+          next unless File.file?(file_name)
+          # What we're doing here is preserving the front matter if we've
+          # already parsed it, and otherwise flagging a file as missing front
+          # matter by creating a nil entry for file_name.
+          file_to_front_matter[file_name] = file_to_front_matter[file_name]
```

Comment:
Okay, the premise makes sense to me.  What's still confusing me about this line is that we come into this method and `file_to_front_matter` starts as an empty hash.  We start iterating over each file in the `_pages` or `pages` directory, and if the file exists and is actually a file, we create/update a hash key referenced by `file_name` and presumably store the value already in there.  But how does any existing value end up in there in the first place since the hash is starting out empty?  Or is it as your comment says, a value of `nil` is going into it?